src/main/java/de/zalando/aruha/nakadi/service/CursorsCommitService.java

```diff
@@ -0,0 +1,120 @@
+import de.zalando.aruha.nakadi.exceptions.NoSuchSubscriptionException;
+import de.zalando.aruha.nakadi.exceptions.ServiceUnavailableException;
+import de.zalando.aruha.nakadi.repository.TopicRepository;
+import de.zalando.aruha.nakadi.repository.db.SubscriptionDbRepository;
+import de.zalando.aruha.nakadi.repository.zookeeper.ZooKeeperHolder;
+import de.zalando.aruha.nakadi.repository.zookeeper.ZooKeeperLockFactory;
+import de.zalando.aruha.nakadi.service.subscription.KafkaClient;
+import de.zalando.aruha.nakadi.service.subscription.SubscriptionKafkaClientFactory;
+import de.zalando.aruha.nakadi.service.subscription.model.Partition;
+import de.zalando.aruha.nakadi.service.subscription.zk.ZkSubscriptionClient;
+import de.zalando.aruha.nakadi.service.subscription.zk.ZkSubscriptionClientFactory;
+
+import java.nio.charset.Charset;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static de.zalando.aruha.nakadi.repository.zookeeper.ZookeeperUtils.runLocked;
+import static java.text.MessageFormat.format;
+
+public class CursorsCommitService {
+
+    private static final Charset CHARSET = Charset.forName("UTF-8");
+
+    private final ZooKeeperHolder zkHolder;
+    private final TopicRepository topicRepository;
+    private final SubscriptionDbRepository subscriptionRepository;
+    private final ZooKeeperLockFactory zkLockFactory;
+    private final ZkSubscriptionClientFactory zkSubscriptionClientFactory;
+    private final SubscriptionKafkaClientFactory subscriptionKafkaClientFactory;
+
+    public CursorsCommitService(final ZooKeeperHolder zkHolder,
+                                final TopicRepository topicRepository,
+                                final SubscriptionDbRepository subscriptionRepository,
+                                final ZooKeeperLockFactory zkLockFactory,
+                                final ZkSubscriptionClientFactory zkSubscriptionClientFactory,
+                                final SubscriptionKafkaClientFactory subscriptionKafkaClientFactory) {
+        this.zkHolder = zkHolder;
+        this.topicRepository = topicRepository;
+        this.subscriptionRepository = subscriptionRepository;
+        this.zkLockFactory = zkLockFactory;
+        this.zkSubscriptionClientFactory = zkSubscriptionClientFactory;
+        this.subscriptionKafkaClientFactory = subscriptionKafkaClientFactory;
+    }
+
+    public boolean commitCursors(final String subscriptionId, final List<Cursor> cursors)
+            throws InvalidCursorException, NoSuchSubscriptionException, ServiceUnavailableException {
+
+        final Subscription subscription = subscriptionRepository.getSubscription(subscriptionId);
+        final String eventType = subscription.getEventTypes().iterator().next();
+
+        createSubscriptionInZkIfNeeded(subscription);
+        topicRepository.validateCommitCursors(eventType, cursors);
+
+        boolean allCommitted = true;
```

Comment:
Maybe you could be a little bit more hipster and use some stream method like `allMatch`, for example?